### ISSUE TITLE
[$250] Chat - Group is created with duplicate user when chat is started quickly on a slow network

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#49931]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**  9.0.41-2
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** biruknew45+1195@gmail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to [https://staging.new.expensify.com/.](https://staging.new.expensify.com/)
2. Open the Developer Tools and go to the Network tab, then set the network speed to 3G
3. Click on the FAB > Start Chat
4. Enter a user's email, then quickly click on it before the fallback avatar changes to the correct user avatar
5. Send a message

## Expected Result:
The chat with the selected user should remain a 1-on-1 chat. The chat should not change into a group with duplicate users

## Actual Result:
After sending the message, the chat turns into a group with duplicate users. It only corrects itself back to a single chat after switching to another chat and then returning to it

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/user-attachments/assets/7ed99952-55d7-44f6-a72f-26f59e044154

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021843049310841478139</li>
        <li>Upwork Job ID: 1843049310841478139</li>
        <li>Last Price Increase: 2024-10-13</li>
<li>Automatic offers: </li>
<ul>
<li>DylanDylann | Contributor | 104420901</li>
<li>nkdengineer | Contributor | 104651259</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @RachCHopkins</details>

**Comment by RachCHopkins** [comment]
I don't understand what's happening in that video. I can't replicate this.

![image](https://github.com/user-attachments/assets/b6e950a8-bd55-4ac9-854d-ec7247d7722d)

I'm on 2 bars of 4G.

Is the person reporting it saying that the two users other than themselves are, in fact, the same user?

**Comment by RachCHopkins** [comment]
Please clarify the "actual result"

**Comment by RachCHopkins** [comment]
Waiting on a response/clarification from QA team.

**Comment by mvtglobally** [comment]
team is still able to reproduce 


https://github.com/user-attachments/assets/44e432c6-24e5-4553-abba-0aab81c6dcb9

**Comment by RachCHopkins** [comment]
Ok, to sum up, it can be reproduced and we don't know what/who the third user is.

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
After sending the message, the chat turns into a group with duplicate users. It only corrects itself back to a single chat after switching to another chat and then returning to it

### What is the root cause of that problem?
If we add a comment when `OpenReport` API is calling, `doesClientNeedToBeUpdated` returns true so the `successData` of `OpenReport` API is added to `deferredUpdates`

https://github.com/Expensify/App/blob/6d8e739f7b1c286ae65565a6689b27b5983fe48a/src/libs/Middleware/SaveResponseInOnyx.ts#L34-L37

<img width="1512" alt="Screenshot 2024-10-07 at 11 53 05" src="https://github.com/user-attachments/assets/f0ab29dd-0611-403a-aded-fb52d29e6788">

After `GetMissingOnyxMessages` API is complete, we detect the pending defer update here but it's empty because `minUpdateID` is not less than the `lastUpdateID` of `OpenReport` API

https://github.com/Expensify/App/blob/6d8e739f7b1c286ae65565a6689b27b5983fe48a/src/libs/actions/OnyxUpdateManager/utils/index.ts#L30

https://github.com/Expensify/App/blob/6d8e739f7b1c286ae65565a6689b27b5983fe48a/src/libs/actions/OnyxUpdateManager/utils/DeferredOnyxUpdates.ts#L42

Then this DM has three members as participant, one is the current user, one is optimistic personal details of the user that we select in the new chat page (it appears because `successData` of `OpenReport` API isn't merged), one is the correct personal of the user.

<img width="1512" alt="Screenshot 2024-10-07 at 12 00 10" src="https://github.com/user-attachments/assets/7827b8b0-bbbb-4bde-b533-6ff91807867b">

### What changes do you think we should make in order to solve the problem?
I think `OpenReport` API doesn't need to wait for any update in other request so we can add `WRITE_COMMANDS.OPEN_REPORT` to `requestsToIgnoreLastUpdateID` array then the `successData` of this API can always be merged and will resolve this issue. We did the same for `WRITE_COMMANDS.DELETE_MONEY_REQUEST,` [here](https://github.com/Expensify/App/issues/49329#issuecomment-2379801120)

```
const requestsToIgnoreLastUpdateID: string[] = [
    WRITE_COMMANDS.OPEN_APP,
    SIDE_EFFECT_REQUEST_COMMANDS.RECONNECT_APP,
    WRITE_COMMANDS.CLOSE_ACCOUNT,
    WRITE_COMMANDS.DELETE_MONEY_REQUEST,
    SIDE_EFFECT_REQUEST_COMMANDS.GET_MISSING_ONYX_MESSAGES,
    WRITE_COMMANDS.OPEN_REPORT,
];
```

https://github.com/Expensify/App/blob/6d8e739f7b1c286ae65565a6689b27b5983fe48a/src/libs/Middleware/SaveResponseInOnyx.ts#L8

### What alternative solutions did you explore? (Optional)

**Comment by c3024** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
On slow networks, if we create a report and comment on it, a group with a duplicate chat is created.
### What is the root cause of that problem?
For write requests, we do not write the response onyxData and the corresponding successData/failureData/finallyData to Onyx directly. Instead, we write them to queuedOnyxUpdates to prevent the replay effect.
https://github.com/Expensify/App/blob/78e484f41e5fb5fa3f8827d2b4aea0ba55b581e4/src/libs/actions/OnyxUpdates.ts#L30-L32
These updates are flushed only after all the PersistedRequests are cleared.
https://github.com/Expensify/App/blob/78e484f41e5fb5fa3f8827d2b4aea0ba55b581e4/src/libs/Network/SequentialQueue.ts#L173-L175
We remove a request from persistent requests only after the response is received and processed with all middlewares.
https://github.com/Expensify/App/blob/78e484f41e5fb5fa3f8827d2b4aea0ba55b581e4/src/libs/Network/SequentialQueue.ts#L90-L99
So, on slow networks, when we call OpenReport, its response and corresponding request’s successData, etc., are added to queuedOnyxUpdates. This cannot be flushed immediately because there was already an AddComment request added to the PersistedRequests.
For some reason that needs to be fixed, the backend sends a previousUpdateID for AddComment that is larger than the previous lastUpdateID/lastClientUpdateID. This makes doesClientNeedToBeUpdated true.
https://github.com/Expensify/App/blob/78e484f41e5fb5fa3f8827d2b4aea0ba55b581e4/src/libs/Middleware/SaveResponseInOnyx.ts#L34
As a result, the AddComment response is not processed directly with OnyxUpdates.apply. It sets this AddComment response to saveUpdateInformation, and the queue is paused. When the queue is paused, the updates waiting in queuedUpdates are not flushed, so OpenReport’s successData waits in the queuedUpdates, and the optimistic user detail is not removed, making the report appear like a group report.
### What changes do you think we should make in order to solve the problem?
1.	First, it needs to be checked why the previousUpdateID for the AddComment response is sent as a higher value than the OpenReport call’s lastUpdatedID/lastClientUpdateID. Fixing this issue should resolve the problem.
2.	However, even with an accidental call to GetMissingOnyxMessages, the previous updates should not be left in limbo without being updated to Onyx until a new write request is made. The updates in the queue need to be written to Onyx after all the missing updates are obtained with GetMissingOnyxMessages. Currently, this is not handled anywhere. When the next write request gets completed and the persistent requests array is empty, all queuedOnyxMessages get updated.
We can see this in the present bug as well: if we send another message, everything updates correctly (the backend also does not send an incorrect previousUpdateID) because the queue is unpaused. After processing this request, when the persisted requests are empty, the queuedOnyxUpdates (which include the previous OpenReport and AddComment responses’ onyxData, successData, etc.) are flushed.
To achieve this, after fetching the missing messages, we should flush the queue finally by changing this:
https://github.com/Expensify/App/blob/78e484f41e5fb5fa3f8827d2b4aea0ba55b581e4/src/libs/actions/OnyxUpdateManager/index.ts#L144
to:
```js
DeferredOnyxUpdates.getMissingOnyxUpdatesQueryPromise()?.finally(finalizeUpdatesAndResumeQueue).then(() => QueuedOnyxUpdates.flushQueue());
```
### What alternative solutions did you explore? (Optional)

**Comment by RachCHopkins** [comment]
@parasharrajat do you like any of the proposals here?

**Comment by DylanDylann** [comment]
I can take over this issue as C+

**Comment by RachCHopkins** [comment]
Oops. I'm not sure if that matters 😬

**Comment by RachCHopkins** [comment]
@DylanDylann how do you feel about the proposals here?

**Comment by DylanDylann** [comment]
on my list today

**Comment by DylanDylann** [comment]
It will take me a bit of time to dive deep into this issue because this seems be complicated

**Comment by c3024** [comment]
@DylanDylann 

Sure!

https://github.com/user-attachments/assets/637e66ac-85b8-4a36-b6c3-f8fc690d061f

1. For all write requests, the backend sends a previousUpdateID and a lastUpdateID. In the above video, the lastUpdateID sent back in the OpenReport response is 2468230213. This means the backend considers this ID to be the last time the client device was updated.
2. Then, when we made an AddComment request, the server’s response returned a previousUpdateID of 2468230219. This means the backend is informing the client that the freshest update it sent to all devices is 2468230219. This is larger than the lastUpdateID (2468230213) saved to the device. The backend is suggesting to the client that there are updates between the OpenReport response (2468230213) and the AddComment response (2468230219). However, this is incorrect because it is impossible for there to be any updates in between, as this is the only client making changes to this account. The backend is sending an incorrect previousUpdateID of 2468230219 when it should be sending the lastUpdateID from the OpenReport response, which is 2468230213.
3. Due to this incorrect previousUpdateID (which is larger than the lastUpdateID), the client thinks it is out of date and pauses applying the AddComment response to Onyx, making a GetOnyxMessages call to retrieve the missing updates. This call returns only the updates between 2468230213 and 2468230219, as can be seen.
4. The lastUpdateID of a write request response should match the previousUpdateID of the next write request if no other updates have been made by another client to the same report, account, etc.

https://github.com/user-attachments/assets/36aa8a34-43f8-4351-9be5-9c3476b29250

5. In the above video, the lastUpdateID of the comment matches the previousUpdateID of the next comment, so no GetOnyxMessages call is made because the client understands that the chain of updates is intact and nothing is missing.

There is a fix required on the backend for the first AddComment response after the OpenReport response. However, there is also a generic problem in the frontend codebase, as mentioned in the RCA and the second point of my suggested solution. The issue is that the last write response (in this case, AddComment, along with the updates waiting in queuedOnyxUpdates) which triggered the GetMissingOnyxMessages call, is not immediately written after the missing updates are fetched though the SequentialQueue is paused after fetching these missed updates. This happens because there is no code to flush the queuedUpdates unless another response is received from the backend, as outlined in the second point of my solution.
>When the next write request gets completed and the persistent requests array is empty, all queuedOnyxMessages get updated.
We can see this in the present bug as well: if we send another message, everything updates correctly (the backend also does not send an incorrect previousUpdateID) because the queue is unpaused. After processing this request, when the persisted requests are empty, the queuedOnyxUpdates (which include the previous OpenReport and AddComment responses’ onyxData, successData, etc.) are flushed.

Feel free to ask if there is still something unclear. As you mentioned, this is a fairly complex issue, and I may have missed explaining some parts fully in an effort to keep the comments concise.

**Comment by DylanDylann** [comment]
@c3024 Thanks for a deep explanation

**Comment by nkdengineer** [comment]
Updated [proposal](https://github.com/Expensify/App/issues/49931#issuecomment-2395918533) 

- Update the RCA and solution

- I suggested the logic [here](https://github.com/Expensify/App/issues/46393#issuecomment-2255598546) but in the past, we thought it was unnecessary and moved the logic to the `flush` function.

**Comment by mkzie2** [comment]
@DylanDylann My solution [here](https://github.com/Expensify/App/issues/51045#issuecomment-2421858831) can fix both issues. @nkdengineer your updated solution is the same as mine.

**Comment by DylanDylann** [comment]
Update: I am on testing solutions

**Comment by DylanDylann** [comment]
@nkdengineer 

> For the reason why GetMissingOnyxMessages API is called I think maybe it's expected from BE.

Why do you think it is expected?

**Comment by DylanDylann** [comment]
Hi @mkzie2, Thanks for your contribution, your solution and @nkdengineer's solution have the same idea. But the @nkdengineer proposed it first in https://github.com/Expensify/App/issues/46393#issuecomment-2255598546

**Comment by DylanDylann** [comment]
Thanks everyone for your inputs

I agree with @c3024 about your RCA

1. `previousUpdateID` on the AddComment API is weird (more detail [here](https://github.com/Expensify/App/issues/49931#issuecomment-2422259161)). But I am not sure about the BE logic. Let's wait for an internal engineer to verify that

2. We don't have the logic to flush OnyxUpdateQueue when it is unpaused. With this problem, I lean toward @nkdengineer's [proposal](https://github.com/Expensify/App/issues/49931#issuecomment-2395918533). 

🎀 👀 🎀 C+ reviewed

**Comment by DylanDylann** [comment]
@c3024 Please give your concern if there is any problem. I really appreciate your investigation with details of RCA

**Comment by c3024** [comment]
That is fine. It is necessary for PersistedRequests to be empty before flushing the queue. While I explained it in the next `AddComment` part, I missed considering it in my solution.

**Comment by aldo-expensify** [comment]
hmm, thanks @c3024 for that detailed explanation: https://github.com/Expensify/App/issues/49931#issuecomment-2422259161

I gave it a first read, I didn't understand 100% because I'm not very familiar with all that, but I think I understood that there is a backend bug related to the onyx updates ids being passed around. If this is true, I'd prefer to prioritize and fix that, and then see again where we are at with this bug.

I'll give it sometime later today to try to reproduce your detailed steps and see if I can understand/see the bug.

**Comment by aldo-expensify** [comment]
> I'll give it sometime later today to try to reproduce your detailed steps and see if I can understand/see the bug.

I fail that day and then I forgot to come back. I'll try to take a better look at this asap

**Comment by DylanDylann** [comment]
not overdue, waiting for @aldo-expensify

**Comment by aldo-expensify** [comment]
Sorry, still haven't had time for this yet.

**Comment by RachCHopkins** [comment]
Waiting on Aldo. Not overdue.

**Comment by aldo-expensify** [comment]
> 1. For all write requests, the backend sends a previousUpdateID and a lastUpdateID. In the above video, the lastUpdateID sent back in the OpenReport response is 2468230213. This means the backend considers this ID to be the last time the client device was updated.
> 2. Then, when we made an AddComment request, the server’s response returned a previousUpdateID of 2468230219. This means the backend is informing the client that the freshest update it sent previously to all devices is 2468230219. This is larger than the lastUpdateID (2468230213) saved to the device. The backend is suggesting to the client that there are updates between the OpenReport response (2468230213) and the AddComment response (2468230219). However, this is incorrect because it is impossible for there to be any updates in between, as this is the only client making changes to this account. The backend is sending an incorrect previousUpdateID of 2468230219 when it should be sending the lastUpdateID from the OpenReport response, which is 2468230213.

I wasn't been able to reproduce this with a slow connection:

<img width="807" alt="image" src="https://github.com/user-attachments/assets/172dcc3d-1fee-4b07-8901-ec89c92bb021">
<img width="743" alt="image" src="https://github.com/user-attachments/assets/85e9cf20-0c5d-414f-bd4c-cda42a4fe3c3">

but then with these steps I was able to reproduce in dev:

1. Go offline
2. Open a chat with a user you haven't chatted before (it can be an account that doesn't exist)
3. Send a message
4. Go online
5. Check the OpenReport's last `lastUpdateID` and the AddComment's `previousUpdateID

I'll investigate what are the updates in between and why these are not matching now.

**Comment by aldo-expensify** [comment]
Hmm... I think the `lastUpdateID` and `previousUpdateID` mismatch come from the `OpenReport` command forwarding a `CreateReportAction` command to create the whisper `"Hi there! This is your first time chatting with..."`. Sometimes the `CreateReportAction` can finish fast enough so its updates are considered in the `previousUpdateID` returned by `OpenReport` and sometimes not.

Here: https://github.com/Expensify/Auth/blob/d3481121749cc204c713c771a13454dbe0165fbe/auth/lib/Report.cpp#L7751

**Comment by aldo-expensify** [comment]
Changing this behaviour in the backend sounds tricky to do. We forward the command to avoid making `OpenReport` faster so there is less DB conflicts.

**Comment by aldo-expensify** [comment]
Assigning @nkdengineer since the proposal to fix the frontend makes sense to me too.

**Comment by DylanDylann** [comment]
@aldo-expensify, Could you help to check [this comment](https://github.com/Expensify/App/pull/51712#issuecomment-2449574667)? I believe that we need to get the PR merged as soon as possible because there are many similar bugs in other places that also be fixed by [our PR](https://github.com/Expensify/App/pull/51712)
- https://github.com/Expensify/App/issues/51819
- https://github.com/Expensify/App/issues/51801 
- https://github.com/Expensify/App/issues/51045

**Comment by aldo-expensify** [comment]
@RachCHopkins I'll be off next week (for the whole week). I don't think this bug is very important, so I think it would be fine to wait until after that. If we don't want to wait, feel free to reassign to another engineer.

The PR https://github.com/Expensify/App/pull/51712 is a bit tricky to review because it is changing some low level behaviour, and I'm not sure if it may cause more harm than good. Someone with more knowledge about the network queue may know better.

**Comment by aldo-expensify** [comment]
I'm making the backend PR ready for review: https://github.com/Expensify/Auth/pull/13168

**Comment by DylanDylann** [comment]
@aldo-expensify @RachCHopkins This issue has already been fixed

https://github.com/user-attachments/assets/75cee946-4fa0-43c7-8297-b98ff8180b27

cc @nkdengineer

**Comment by DylanDylann** [comment]
Other issues that have the same RCA have also been fixed

https://github.com/Expensify/App/issues/51801#issuecomment-2527163900
https://github.com/Expensify/App/issues/51045#issuecomment-2527152180

**Comment by DylanDylann** [comment]
@nkdengineer This PR seems unnecessary. I think we can close it

**Comment by tgolen** [comment]
I closed the PR this morning. I think this issue can be resolved, but @nkdengineer did a good amount of work on this, so I think the bounty should be reduced by 50% to $125 and still pay out for this @RachCHopkins

**Comment by RachCHopkins** [comment]
Sounds good to me @tgolen !

**Comment by RachCHopkins** [comment]
Payment Summary:

- Contributor: @nkdengineer paid $125 via Upwork
- Contributor+: @DylanDylann paid $125 via Upwork

Upwork job [here](https://www.upwork.com/jobs/~021843049310841478139)

**Comment by RachCHopkins** [comment]
Contributor has been paid, the contract has been completed, and the Upwork post has been closed.

**Comment by DylanDylann** [comment]
@RachCHopkins As from [our guideline](https://expensify.slack.com/archives/C02NK2DQWUX/p1723055151068559?thread_ts=1723015261.405329&cid=C02NK2DQWUX), the contributor should get full compensation in this case. Could you please evaluate it again?

**Comment by aldo-expensify** [comment]
Reopening so this gets an answer: https://github.com/Expensify/App/issues/49931#issuecomment-2562137152

**Comment by RachCHopkins** [comment]
Sorry @aldo-expensify @DylanDylann @tgolen I'm not understanding who should be paid what here. 

@tgolen notes above that it should be a 50% payment because some work was done, but never implemented. 

@DylanDylann has linked to a post that says that when someone has a solution which is implemented but later reverted and then not used moving forward due to internal Expensify BE work not being complete, they should get a full payment. 

I see in this case that @nkdengineer's proposal was accepted, but not implemented. Which is different to the case discussed above (but please correct me if I'm wrong), and this whole issue was deemed unnecessary because everything was fixed elsewhere.

**Comment by DylanDylann** [comment]
@RachCHopkins From https://expensify.slack.com/archives/C02NK2DQWUX/p1723055151068559?thread_ts=1723015261.405329&cid=C02NK2DQWUX, @mallenexpensify said that:

> If the deliverable has changed because the bug doesn't exist anymore, another PR has fixed it, or we've decided it's no longer a priority to fix, pay the contributor and C+ 100% of the job price (they were hired to do a job and did the work)

In this issue, contributors are already hired and the PR is also ready (almost work has already been done). But the deliverable has changed because we decided to fix it on the BE. So I think the contributors should get fully paid

Anyway, I will only note that in case we overlook the [BZ SO](https://stackoverflowteams.com/c/expensify/questions/8719). If the down price is intentional, let's close this one.

**Comment by RachCHopkins** [comment]
I'm just the girl with the checkbook here, so will wait on @tgolen and @aldo-expensify

**Comment by tgolen** [comment]
OK, I didn't know about that SO. The policy is clear in the SO, so let's just follow it and pay out 100%.

**Comment by mallenexpensify** [comment]
@DylanDylann thanks for the link, tag and quote text.  That _might_ be a lil outdated. Or, the below, in the [main payment SO](https://stackoverflowteams.com/c/expensify/questions/7582/7583#7583), might be more updated.   

> If a contributor has been hired for a job and we decide to close the job before it is successfully completed
> 50% payment to contributor and C+ if the contributor was 🎀👀🎀 by a C+ or hired/assigned.
> 
> 100% due if contributor drafted a PR and requested a review.
> 
> 100% due to C+ if they reviewed the PR.
> 
> One caveat here might be if the hired contributor wasn't working on their PR with urgency.

Reasoning.. I don't like the idea that someone could be hired, do no work on the PR then get full payment because _something_ happened between the time they were hired and they submitted their PR, we don't want to encourage people to move slowly with a possible payout.  (that def doesn't seem to be what happened here, just providing an example for the process). 

Based on the above process, it appears 100% compensation is due based on @nkdengineer 's [PR](https://github.com/Expensify/App/pull/51712).

**Comment by DylanDylann** [comment]
This is a struggle with many challenges while working on it. In this issue, I also worked on reviewing [the PR](https://github.com/Expensify/App/pull/51712) when it was created and discovered [another problem](https://github.com/Expensify/App/pull/51712#issuecomment-2449574667) that started a long discussion and end up with the fix on the BE side. After the fix is deployed I also help to verify all [related issues](https://github.com/Expensify/App/issues/49931#issuecomment-2527170612) are fixed

I raised this concern because I don't know the exact process in this case. @mallenexpensify Thanks for the official information.

**Comment by mallenexpensify** [comment]
@DylanDylann please always reference past posts and details.  We appreciate the help and the all the invested time.  My goal is to keep processes up to date, to make them fair and to discuss with contributors if they don't agree.

**Comment by RachCHopkins** [comment]
Ok, so I paid out the Contributor and C+ $125 each here. So it looks like I need to go back and do some remedial work as far as payments go. I shall work out the best way to do so!
